### PR TITLE
Object reconstruction in Photon

### PR DESCRIPTION
**Pull request by stephanie-wang**
This exposes object reconstruction to the photon client. On a request for object reconstruction from the client, the photon local scheduler will try to resubmit the task that created the object for execution.

### CODE REVIEW

**Review by pcmoritz — APPROVED** [review]

File: src/photon/test/photon_tests.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    }
+    /* Make sure we receive each task from the initial submission. */
+    for (int i = 0; i < NUM_TASKS; ++i) {
+      task_spec *task_assigned = photon_get_task(photon->conn);
+      ASSERT_EQ(memcmp(task_assigned, specs[i], task_spec_size(task_assigned)),
+                0);
+      free_task_spec(task_assigned);
+    }
+    /* Request reconstruction of the last return object. */
+    object_id return_id = task_return(specs[9], 0);
```
Comment (robertnishihara) [inline src/photon/test/photon_tests.c]: `NUM_TASKS - 1` instead of 9

File: src/photon/photon_scheduler.h [hunk 2]
```diff
@@ -50,4 +50,34 @@ void process_plasma_notification(event_loop *loop,
                                  void *context,
                                  int events);
 
+/**
+ * Reconstruct an object. This triggers a single reexecution of the task that
+ * originally created the object, even if the object does exist on some node.
```
Comment (robertnishihara) [inline src/photon/photon_scheduler.h]: If the object table says that the object is present on some nodes, then actually this does not trigger any reexecution, right?

File: src/photon/test/photon_tests.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+const char *plasma_socket_name_format = "/tmp/plasma_socket_%d";
+const char *photon_socket_name_format = "/tmp/photon_socket_%d";
+
+int64_t timeout_handler(event_loop *loop, int64_t id, void *context) {
+  event_loop_stop(loop);
+  return EVENT_LOOP_TIMER_DONE;
+}
+
+typedef struct {
+  /* A socket to mock the Plasma store. */
```
Comment (robertnishihara) [inline src/photon/test/photon_tests.c]: I don't feel that strongly about this case, but for structs we should probably document with `/** */`

**Review by robertnishihara — COMMENTED** [review]
Nice job!

**Pull request merged**